modules/templates/eval/eval.go

```diff
@@ -0,0 +1,332 @@
+		return Num{truth(n1 != n2)}
+	case "<":
+		return Num{truth(n1 < n2)}
+	case "<=":
+		return Num{truth(n1 <= n2)}
+	case ">":
+		return Num{truth(n1 > n2)}
+	case ">=":
+		return Num{truth(n1 >= n2)}
+	case "and":
+		t1, _ := util.ToFloat64(n1)
+		t2, _ := util.ToFloat64(n2)
+		return Num{truth(t1 != 0 && t2 != 0)}
+	case "or":
+		t1, _ := util.ToFloat64(n1)
+		t2, _ := util.ToFloat64(n2)
+		return Num{truth(t1 != 0 || t2 != 0)}
+	}
+	panic("unknown operator: " + string(op))
+}
+
+func applyOp2(op operator, n1, n2 Num) Num {
+	float := false
+	if _, ok := n1.Value.(float64); ok {
+		float = true
+	} else if _, ok = n2.Value.(float64); ok {
+		float = true
+	}
+	if float {
+		f1, _ := util.ToFloat64(n1.Value)
+		f2, _ := util.ToFloat64(n2.Value)
+		return applyOp2Generic(op, f1, f2)
+	}
+	return applyOp2Generic(op, n1.Value.(int64), n2.Value.(int64))
+}
+
+func toOp(v any) (operator, error) {
+	if v, ok := v.(string); ok {
+		return operator(v), nil
+	}
+	return "", fmt.Errorf("unable to convert %q to operator", v)
+}
+
+func (op operator) hasOpenBracket() bool {
+	return strings.HasSuffix(string(op), "(")
+}
+
+func (op operator) isComma() bool {
+	return op == ","
+}
+
+func (op operator) isCloseBracket() bool {
+	return op == ")"
+}
+
+type ExprError struct {
+	msg    string
+	tokens []any
+	err    error
+}
+
+func (err ExprError) Error() string {
+	sb := strings.Builder{}
+	sb.WriteString(err.msg)
+	sb.WriteString(" [ ")
+	for _, token := range err.tokens {
+		_, _ = fmt.Fprintf(&sb, `"%v" `, token)
+	}
+	sb.WriteString("]")
+	if err.err != nil {
+		sb.WriteString(": ")
+		sb.WriteString(err.err.Error())
+	}
+	return sb.String()
+}
+
+func (err ExprError) Unwrap() error {
+	return err.err
+}
+
+func (e *Eval) applyOp() {
+	op := e.stackOp.pop()
+	if op == "not" {
+		num := e.stackNum.pop()
+		i, _ := util.ToInt64(num.Value)
+		e.stackNum.push(Num{truth(i == 0)})
+	} else {
+		num2 := e.stackNum.pop()
+		num1 := e.stackNum.pop()
+		e.stackNum.push(applyOp2(op, num1, num2))
+	}
+}
+
+// Exec evaluates the given expression tokens and returns the result.
+// If no error occurs, the result is either an int64 or a float64.
+// If all numbers are integer, the result is an int64, otherwise if there is any float number, the result is a float64.
+// Golang's template syntax supports comparable int types: {{if lt $i32 $i64}} is right.
+func (e *Eval) Exec(tokens ...any) (ret Num, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			rErr, ok := r.(error)
+			if !ok {
+				rErr = fmt.Errorf("%v", r)
+			}
+			err = ExprError{"invalid expression", tokens, rErr}
+		}
+	}()
+	for _, token := range tokens {
+		n, err := toNum(token)
+		if err == nil {
+			e.stackNum.push(n)
+			continue
+		}
+
+		op, err := toOp(token)
+		if err != nil {
+			return Num{}, ExprError{"invalid expression", tokens, err}
+		}
+
+		switch {
+		case op.hasOpenBracket():
+			e.stackOp.push(op)
+		case op.isCloseBracket(), op.isComma():
+			var sop operator
```

Comment:
`sop`?
What's that supposed to be?
`specialOp`?

Reply:
stack op